src/xmerl_dsig.erl

```diff
@@ -0,0 +1,731 @@
+    {KeyFun, CertBin}.
+
+
+sign_and_verify_test() ->
+    {Doc, _} = xmerl_scan:string("<?xml version=\"1.0\" encoding=\"utf-8\"?><x:foo id=\"test\" xmlns:x=\"urn:foo:x:\"><x:name>blah</x:name><ds:Signature xmlns:ds=\"http://www.w3.org/2000/09/xmldsig#\"><ds:SignedInfo><ds:CanonicalizationMethod Algorithm=\"http://www.w3.org/2001/10/xml-exc-c14n#\"/><ds:SignatureMethod Algorithm=\"http://www.w3.org/2000/09/xmldsig#rsa-sha1\"/><ds:Reference URI=\"\"><ds:Transforms><ds:Transform Algorithm=\"http://www.w3.org/2000/09/xmldsig#enveloped-signature\"/></ds:Transforms><ds:DigestMethod Algorithm=\"http://www.w3.org/2000/09/xmldsig#sha1\"/><ds:DigestValue></ds:DigestValue></ds:Reference></ds:SignedInfo><ds:SignatureValue></ds:SignatureValue><ds:KeyInfo><ds:X509Data><ds:X509SubjectName></ds:X509SubjectName><ds:X509Certificate></ds:X509Certificate></ds:X509Data></ds:KeyInfo></ds:Signature></x:foo>", [{namespace_conformant, true}]),
+    {Key, CertBin} = test_sign_key(),
+    SignedXml = sign(Doc, Key, CertBin),
+    ok = verify(SignedXml, [crypto:hash(sha, CertBin)]).
+
+sign_and_verify_sha256_test() ->
+    {Doc, _} = xmerl_scan:string("<?xml version=\"1.0\" encoding=\"utf-8\"?><x:foo id=\"test\" xmlns:x=\"urn:foo:x:\"><x:name>blah</x:name><ds:Signature xmlns:ds=\"http://www.w3.org/2000/09/xmldsig#\"><ds:SignedInfo><ds:CanonicalizationMethod Algorithm=\"http://www.w3.org/2001/10/xml-exc-c14n#\"/><ds:SignatureMethod Algorithm=\"http://www.w3.org/2000/09/xmldsig#rsa-sha1\"/><ds:Reference URI=\"\"><ds:Transforms><ds:Transform Algorithm=\"http://www.w3.org/2000/09/xmldsig#enveloped-signature\"/></ds:Transforms><ds:DigestMethod Algorithm=\"http://www.w3.org/2000/09/xmldsig#sha1\"/><ds:DigestValue></ds:DigestValue></ds:Reference></ds:SignedInfo><ds:SignatureValue></ds:SignatureValue><ds:KeyInfo><ds:X509Data><ds:X509SubjectName></ds:X509SubjectName><ds:X509Certificate></ds:X509Certificate></ds:X509Data></ds:KeyInfo></ds:Signature></x:foo>", [{namespace_conformant, true}]),
+    {Key, CertBin} = test_sign_256_key(),
+    SignedXml = sign(Doc, Key, CertBin),
+    ok = verify(SignedXml, [crypto:hash(sha, CertBin)]).
+
+%% We don't generate id's in this implementation
+%% sign_generate_id_test() ->
+%%     {Doc, _} = xmerl_scan:string("<x:foo xmlns:x=\"urn:foo:x:\"><x:name>blah</x:name></x:foo>", [{namespace_conformant, true}]),
+%%     {Key, CertBin} = test_sign_key(),
+%%     SignedXml = sign(Doc, Key, CertBin),
+%%     Ns = [{"ds", 'http://www.w3.org/2000/09/xmldsig#'}],
+%%     [#xmlAttribute{name = 'ID', value = RootId}] = xmerl_xpath:string("@ID", SignedXml, [{namespace, Ns}]),
+%%     [#xmlAttribute{value = "#" ++ RootId}] = xmerl_xpath:string("ds:Signature/ds:SignedInfo/ds:Reference/@URI", SignedXml, [{namespace, Ns}]).
+
+utf8_test() ->
+    Name = <<208,152,208,179,208,190,209,128,209,140,32,208,154,
+             208,176,209,128,209,139,208,188,208,190,208,178,32>>,
+    ThisPerson = <<227,129,157,227,129,174,228,186,186,10>>,
```

Comment:
This would be much more readable if the unicode strings were used:

```erl
    Name = <<"Игорь Карымов "/utf8>>.,
     ThisPerson = <<"その人\n"/utf8>>,
```

Reply:
agreed.